src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,143 @@
+/**
+ * @typedef {(request: Request) => Promise<Response>} MiddlewareFunc
+ * @typedef {Parameters<ConstructorParameters<typeof Promise>[0]>} PromiseCallbackParams
+ * @typedef {{resolve: PromiseCallbackParams[0]; reject: PromiseCallbackParams[1]}} Resolvers
+ * @typedef {{resolvers: Resolvers, removeAbortListener?: Function}} DeduplicatedRequest
+ *
+ * @typedef {object} InflightRequestEntry
+ * @prop {Request} inflightRequest
+ * @prop {AbortController | null} abortController
+ * @prop {Record<string, DeduplicatedRequest>} dedupedRequests
+ *
+ * @typedef {Record<string, InflightRequestEntry>} InflightRequestInfo
+ */
+
 export class D2LFetchDedupe {
 
 	constructor() {
-		this._inflightRequests = this._inflightRequests || [];
+		/** @type {InflightRequestInfo} */
+		this._inflightRequests = {};
+		this._nextReqId = 0;
 	}
 
+	/**
+	 * @param {Request} request
+	 * @param {MiddlewareFunc} next
+	 *
+	 * @returns {Promise<Response>}
+	 */
 	dedupe(request, next) {
 		if (false === request instanceof Request) {
 			return Promise.reject(new TypeError('Invalid request argument supplied; must be a valid window.Request object.'));
 		}
 
 		if (request.method !== 'GET' && request.method !== 'HEAD' && request.method !== 'OPTIONS') {
-			if (!next) {
-				return Promise.resolve(request);
-			}
 			return next(request);
 		}
 
 		const key = this._getKey(request);
-		if (this._inflightRequests[key]) {
-			this._inflightRequests[key].count++;
-			return this._inflightRequests[key].action;
-		}
+		const reqId = this._nextReqId++;
+
+		let newRequest = false;
+
+		if (!this._inflightRequests[key]) {
+			const abortController = window.AbortController
+				? new AbortController()
+				: null;
```

Comment:
If the request that is initiating this isn't using an abortController, then there's at least one request that isn't abortable, and as such the request as a whole won't be abortable, and we don't need to new up this abortcontroller.

Does that sound right?

```suggestion
			const abortController = request.signal && typeof request.signal.addEventListener === 'function' && window.AbortController
				? new AbortController()
				: null;
```

Reply:
In the general case (`AbortController` is natively supported), a request will always have an `AbortSignal`, even if not explicitly provided, so checking for the constructor function is equivalent. That said, at least some fetch polyfills don't match this by constructing default signals if available, so still seems reasonable.

Reply:
Ah okay. I was thinking along the lines of "don't new up an abort controller, if the incoming request wasn't using an abort controller, since we'll never use it", but if we can't detect that, then :shrug: